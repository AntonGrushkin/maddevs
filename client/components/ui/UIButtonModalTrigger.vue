<template>
  <button
    class="ui-button-modal-trigger"
    :class="{ 'ui-button-modal-trigger--red': isRed, 'ui-button-modal-trigger--black': isBlack, 'ui-button-modal-trigger--grey': isGrey }"
    @click="showModal()"
    v-WaveAnimation
  >
    {{ buttonInnerText }}
  </button>
</template>

<script>
import WaveAnimation from '@/directives/WaveAnimation';

export default {
  name: 'UIButtonModalTrigger',
  directives: {
    WaveAnimation
  },
  props: {
    buttonInnerText: {
      type: String,
      default: 'Button Inner Text'
    },
    modalWindowName: {
      type: String,
      default: ''
    },
    isRed: {
      type: Boolean,
      default: false
    },
    isBlack: {
      type: Boolean,
      default: false
    },
    isGrey: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      scrollYPosition: null
    };
  },
  mounted() {
    window.addEventListener('scroll', this.handleScroll);
  },
  destroyed() {
    window.removeEventListener('scroll', this.handleScroll);
  },
  methods: {
    showModal() {
      this.$modal.show(this.$props.modalWindowName);
      this.disableScrollOnBody();
    },
    disableScrollOnBody() {
      if (window.innerWidth <= 640) {
        document.body.style.position = 'fixed';
      }
      document.body.style.top = `-${this.scrollYPosition}`;
      document.body.style.overflow = 'hidden';
    },
    handleScroll() {
      this.scrollYPosition = `${window.scrollY}px`;
    }
  }
};
</script>

<style lang="scss" scoped>
  @import '../../assets/styles/vars';

  .ui-button-modal-trigger {
    height: 65px;
    border: 1px solid $border-color--grey;
    border-radius: 2px;
    box-shadow: none;
    background-color: transparent;
    font-size: 18px;
    font-family: 'Poppins-Regular', sans-serif;;
    cursor: pointer;
    transition: 0.4s;

    &--red {
      color: $text-color--red;
      border: 1px solid $border-color--red;

      &:hover {
        background-color: $bgcolor--red;
        color: $text-color--white;
      }
    }

    &--black {
      color: $text-color--black;
      border: 1px solid $border-color--black;

      &:hover {
        background-color: $bgcolor--black;
        color: $text-color--white;
      }
    }

    &--grey {
      color: $text-color--grey;
      border: 1px solid $border-color--grey;

      &:hover {
        background-color: $bgcolor--grey;
        color: $text-color--black;
      }
    }

    &:active {
      background-color: $button-active--red;
      border-color: $button-active-border--red;
    }
  }
</style>
