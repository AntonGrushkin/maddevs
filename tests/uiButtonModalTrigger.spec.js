import {
  mount
} from '@vue/test-utils';
import UIButtonModalTrigger from '@/components/ui/UIButtonModalTrigger';

describe('Ui button modal trigger', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(UIButtonModalTrigger, {
      propsData: {
        buttonInnerText: 'Some button name',
        modalWindowName: 'Some modal window name',
        isRed: true
      },
      mocks: {
        $modal: {
          show: jest.fn()
        },
        $nuxt: {
          $emit: jest.fn()
        }
      }
    });
  });

  // ------ IMPORTANT ----- //
  test('is a Vue instance', () => {
    expect(wrapper.exists()).toBeTruthy();
  });
  
  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
  // --------------------- //

  test('correctly props data', () => {
    expect(wrapper.props().buttonInnerText).toBe('Some button name');
    expect(wrapper.props().modalWindowName).toBe('Some modal window name');
  });

  test('sets the correctly button name and contains new class --red', () => {
    let button = wrapper.find('.ui-button-modal-trigger');
    expect(button.text()).toBe('Some button name');
    expect(button.classes()).toContain('ui-button-modal-trigger--red');
  });

  test('Should call two events', () => {
    wrapper.vm.showModal();
    expect(wrapper.vm.$modal.show).toHaveBeenCalled();
    expect(wrapper.vm.$nuxt.$emit).toHaveBeenCalled();
  });
});
