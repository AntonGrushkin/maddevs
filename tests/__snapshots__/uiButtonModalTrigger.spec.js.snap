// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ui button modal trigger renders correctly 1`] = `
<button
  class="ui-button-modal-trigger ui-button-modal-trigger--red"
  style="position: relative; overflow: hidden;"
>
  
  Some button name

  <div
    class="wave"
  />
</button>
`;
